internal/log/slog.go

```diff
@@ -0,0 +1,107 @@
+// Copyright 2021 The prometheus-operator Authors
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+package log
+
+import (
+	"fmt"
+	"log/slog"
+	"os"
+	"strings"
+	"time"
+)
+
+// NewLoggerSlog returns a *slog.Logger that prints in the provided format at the
+// provided level with a UTC timestamp and the caller of the log entry.
+func NewLoggerSlog(c Config) (*slog.Logger, error) {
+	lvlOption, err := parseLevel(c.Level)
+	if err != nil {
+		return nil, err
+	}
+
+	logger, err := parseFmt(lvlOption, c.Format)
+	if err != nil {
+		return nil, err
+	}
+
+	return logger, nil
+}
+
+func replaceSlogAttributes(_ []string, a slog.Attr) slog.Attr {
+	if a.Key == "time" {
+		return slog.Attr{
+			Key:   "ts",
+			Value: slog.StringValue(a.Value.Time().UTC().Format(time.RFC3339Nano)),
+		}
+	}
+
+	if a.Key == "level" {
+		return slog.Attr{
+			Key:   "level",
+			Value: slog.StringValue(strings.ToLower(a.Value.String())),
+		}
+	}
+
+	if a.Key == "source" {
+		return slog.Attr{
+			Key:   "caller",
+			Value: a.Value,
+		}
+	}
+
+	return a
+}
+
+func parseFmt(lvlOption slog.Level, format string) (*slog.Logger, error) {
```

Comment:
IMO this functions is doing more than parsing the format, it's also initializing the logger with the proper handler.

Maybe you can have one approach like this.

```go
func getHandlerFromFormat(format string, opts slog.HandlerOptions) (slog.Handler, error) {
    var handler slog.Handler
    switch {
	case logFormat == "text":
		handler = slog.NewTextHandler(os.Stdout, handlerOptions)
	case logFormat == "json":
		handler = slog.NewJSONHandler(os.Stdout, handlerOptions)
	default:
		return nil, errors.New("unknown log format")
	}
}
```

So you can change your `NewLoggerSlog` to looks like this:

```go
func NewLoggerSlog(c Config) (*slog.Logger, error) {
	lvlOption, err := parseLevel(c.Level)
	if err != nil {
		return nil, err
	}

	handler, err := getHandlerFromFormat(lvlOption, slog.HandlerOptions{
			Level:       lvlOption,
			AddSource:   true,
			ReplaceAttr: replaceSlogAttributes,
		})
	if err != nil {
		return nil, err
	}

	return slog.New(h), nil
}
```

wdyt?

Reply:
yep, makes sense, will also simplify the tests. Adressing

Reply:
addressed it here -> https://github.com/prometheus-operator/prometheus-operator/pull/6743/commits/b9ea62456d78c47b27005fbeb945bd3f202d8377